### PR TITLE
Changes to support microscopy viewer to call dcmjs in order to get the report format from ROI

### PR DESCRIPTION
**Pull request by biharck**
### What
Refactoring from Polygon to Polyline
gracefully failure when a feature is undefined
fixing getAllROIs and refactoring getROI
adding z coord as 0 and adding geometry name
returning coordinateFormatFunction at getROI function

### Why
Changes to support microscopy viewer to call dcmjs in order to get the report format from ROI

### CODE REVIEW

File: src/api.js [hunk 1]
```diff
@@ -100,14 +101,37 @@ function _scoord2Geometry(scoord) {
 
 function _geometryCoordinates2scoordCoordinates(coordinates) {
   // TODO: Transform to coordinates on pyramid base layer???
-  return [coordinates[0] + 1, -coordinates[1]]
+  return [coordinates[0] + 1, -coordinates[1], 0]
```
Comment (swederik) [inline src/api.js]: I think the conversion to SCOORD3D should be in dcmjs, so we don't need this third argument here.

File: src/api.js [hunk 2]
```diff
@@ -596,32 +610,39 @@ class VLWholeSlideMicroscopyImageViewer {
...
         type: 'Circle',
+        geometryName: 'Circle'
       },
       box: {
         type: 'Circle',
+        geometryName: 'Box',
         geometryFunction: createRegularPolygon(4),
       },
-      polygon: {
+      polyline: {
```
Comment (swederik) [inline src/api.js]: I don't think we need to change the property name

File: yarn.lock [hunk 3]
```diff
@@ -0,0 +1,3261 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (swederik) [inline yarn.lock]: We already have package-lock, so we don't need this file (you can add to gitignore I guess)

File: src/api.js [hunk 4]
```diff
@@ -37,7 +37,8 @@ import DICOMwebClient from 'dicomweb-client/build/dicomweb-client.js'
 
 
 function _geometry2Scoord(geometry) {
-  const type = geometry.getType()
+  // const name = geometry.constructor.name;
```
Comment (swederik) [inline src/api.js]: I guess we don't need this line

File: yarn.lock (hunk 3 above)
Comment (biharck) [reply yarn.lock]: done

File: src/api.js (hunk 2 above)
Comment (biharck) [reply src/api.js]: done

File: src/api.js (hunk 4 above)
Comment (biharck) [reply src/api.js]: done

File: src/api.js (hunk 1 above)
Comment (biharck) [reply src/api.js]: Done

**Comment by hackermd** [comment]
@biharck thanks for your contributions! Below are a few comments on the viewer interface and the specification of spatial coordinates.

Don't set `coordinateSystem` property on `ROI` objects. Rather use a `scoord` of a different type:

- `Scoord3D` for coordinates in millimeter unit defined relative to the slide coordinate system
- `Scoord` for coordinates in pixel unit defined relative to the total pixel matrix

However, I realized that there are some issues with using *SCOORD* for *VL Whole Slide Microscopy Image* instances. Therefore, we should (at least for now) only use *SCOORD3D*. This also makes more sense conceptually, since the ROIs are vector graphics.

Accordingly, classes defined in `scoord.js` need to be refactored (see http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.9.html):

1. base class `Scoord` should be renamed to `Scoord3D`
2. getter `pixelOriginInterpretation` needs to be removed and a new property `referendedFrameOfReferenceUID` needs to be added (see http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.9.html#para_14610131-0a52-4367-8ebe-36688ff8305d)
4. a `Polygon` class should be implemented (in contrast to *SCOORD*, *SCOORD3D* supports graphic type *POLYGON* (http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.9.html#sect_C.18.9.1.2)

Ultimately, we may want to move `roi.js` and `scoord.js` into the *dcmjs* library.

**Comment by hackermd** [comment]
Following the previous comment, I would remove the `coordinateSystem` parameter of function `_getROIByFeature()` and of methods `getAllROIs()` and `getROI()` of class `VLWholeSlideMicroscopyImageViewer`.

If you prefer to keep a parameter (in case we may later reintroduce support for *SCOORD*), I would name the parameter `unit` (with choices `"mm"`/`"millimeter"` and `"pixel"`) and raise an error `if (unit === "pixel")` with a message to indicate that *SCOORD* is not implemented.

**Pull request merged**